### PR TITLE
Sequence to Sequence Problem Type

### PR DESCRIPTION
**Pull request by psinger**
Closes https://github.com/h2oai/h2o-llmstudio/issues/100

Adds a new problem Type Sequence to Sequence Modeling. For now tested for different T5 models.
We can probably go step by step adding more functionality if this will be frequently used.

I tried to reuse as much as possible to not have much duplicate code.

@maxjeblick please take an initial look, it might need 1-2 iterations

Todo:

- [x] Rework dataset?
- [x] Check padding
- [x] Check merging / pipeline
- [x] Default settings for dtype etc?
- [x] Model card

### CODE REVIEW

**Review by maxjeblick — COMMENTED** [review]
Thanks a lot for adding this problem type, it looks in a very good state!

I would move `prepare_lora` and `generate` to separate functions to avoid code duplication (also, they will probably be needed for other potential problem types as well).
Apart from that, the code looks fine, will test it more the next days.

**Comment by psinger** [comment]
Thx for the comments @maxjeblick - tried to address them.

File: llm_studio/src/utils/config_utils.py [hunk 1]
```diff
@@ -204,6 +204,10 @@ def load_config_yaml(path: str):
         from llm_studio.python_configs.text_causal_language_modeling_config import (
             ConfigProblemBase,
         )
+    elif cfg_dict["problem_type"] == "text_sequence_to_sequence_modeling":
```
Comment (maxjeblick) [inline llm_studio/src/utils/config_utils.py:207]:
```
def load_config_yaml(path: str):
    """Loads config from yaml file

    Args:
        path: path of file to load from
    Returns:
        config object
    """
    with open(path, "r") as fp:
        cfg_dict = yaml.load(fp, Loader=yaml.FullLoader)

    problem_type = cfg_dict["problem_type"]
    module_name = f"llm_studio.python_configs.{problem_type}_config"
    try:
        module = importlib.import_module(module_name)
    except ModuleNotFoundError:
        raise NotImplementedError(f"Problem Type {problem_type} not implemented")
    return module.ConfigProblemBase.from_dict(cfg_dict)

```

This code snippet should work, may useful to switch to easy subsequent adding of new problem types.

**Review by maxjeblick — CHANGES_REQUESTED** [review]
Thanks for the changes! Some observations:

- Summary tab fails with `AttributeError: 'ConfigProblemBase' object has no attribute 'hf'`
- Train Data Insights currently shows decoded input ids, whereas the model uses prompt input ids. Maybe it is useful in general to show prompt input ids, instead of input ids? Could potentially also changed in #301.
- Mid-term, promoting `generate` and `prepare_lora` to dedicated functions (with corresponding arguments, `cfg`, `backbone`, ...) may be useful. We can leave it here as is and I can have a look in #317.

File: llm_studio/src/utils/config_utils.py (hunk 1 above)
Comment (maxjeblick) [reply llm_studio/src/utils/config_utils.py:207]: You also leave as is, as it is already part of RLHF PR

**Comment by maxjeblick** [comment]
The PR is probably good to merge after hf has been added to the config.

**Comment by psinger** [comment]
Yes, I just noticed that I also need to adjust the summary model card now. On it.

**Review by maxjeblick — APPROVED** [review]
LGTM thanks!

One small issue:
`trust_remote_code` is not populated in the model card, should be easy to fix.

**Pull request merged**